---
title: "Sequence assembly"
exercises: 60
questions:
- "How can the information in the sequencing reads be reduced?"
- "What are the different methods for assembly?"
objectives:
- "Understand differences between assembly methods"
- "Assemble the short reads"
- "What is the effect of different k-mer sizes"
keypoints:
- "Assembly is a process which aligns and merges fragments from a longer DNA sequence in order to reconstruct the original sequence."
- "k-mers are short fragments of DNA of length k"
---

Sequence assembly means the alignment and merging of reads in order to reconstruct the original sequence. The assembly of a genome from short sequencing reads will take a while - from minutes up to several hours per genome. 

## Sequence assembly

The assembler we will run is [SPAdes](http://cab.spbu.ru/software/spades/). SPAdes generates a final assembly from multiple kmers. A list of kmers is automatically selected by SPAdes using the maximum read length of the input data, and each individual kmer contributes to the final assembly. If desired, a list of kmers can be specified with the -k flag which will override automatic kmer selection.


## Preparation

To run SPAdes we will use the spades.py command with the --only-assembler option as the reads have already been trimmed, -o for the output folder, -1 for the path to the forward reads, -2 for the path to the reverse reads. To speed up the process, we will be using only one k-mer size of 55 basepair.

~~~
$ spades.py -1 ERR022075_1.trimmed.fastq -2 ERR022075_2.trimmed.fastq -k 55 -o ERR022075
$ cd ~/assembly
$ ls 
$ cd ERR026473
$ ls
~~~
{: .bash}

The assemblies for every k-mer setting are found in the folders marked K21, K33 K55 and are called final_contigs.fasta. The end result before scaffolding is called contigs.fasta and the scaffolded contigs are available in the file called scaffolds.fasta.

> ## Challenge: How many contigs were generated by SPAdes?
>
> Find out how many contigs or scaffolds there are in the *E. coli* isolate. 
>
> Hint:
> ~~~
> $ grep -c
> ~~~
> prints a count of matching lines for each input file.
> 
> > ## Solution
> >
> > 
> > ~~~
> > $ grep -c '>'  ERR*/K55/final_contigs.fasta
> > $ grep -c '>'  ERR*/contigs.fasta
> > $ grep -c '>'  ERR*/scaffolds.fasta
> > 
> > ~~~
> > {: .output}
> {: .solution}
{: .challenge}



{% include links.md %}
