---
title: "Sequence assembly"
exercises: 60
questions:
- "How can the information in the sequencing reads be reduced?"
- "What is the effect of k-mer size on assembly?"
objectives:
- "Assemble the short reads"
keypoints:
- "Assembly is a process which aligns and merges fragments from a longer DNA sequence in order to reconstruct the original sequence."
- "k-mers are short fragments of DNA of length k"
---

Sequence assembly means the alignment and merging of reads in order to reconstruct the original sequence. The assembly of a genome from short sequencing reads will take a while - from minutes up to several hours per genome. 

## Sequence assembly

The assembler we will run is [SKESA](https://github.com/ncbi/SKESA). SKESA is a DeBruijn graph-based de-novo assembler designed for assembling reads of microbial genomes sequenced using Illumina. It uses multiple k-mer lengths during assembly. A list of kmers is automatically selected by SKESA using the maximum read length of the input data, and each individual kmer contributes to the final assembly. In the file allkmers.txt we will collect all contigs produced at different k-mer sizes.


~~~
$ cd ~/Secretome_prediction/trimmed
$ skesa --cores 2 --memory 32 --all allkmers.txt --fastq ERR022075_1.trimmed.fastq ERR022075_2.trimmed.fastq > ~/Secretome_prediction/assembly/ERR022075.fasta
~~~
{: .bash}

This will take a while. Try not to interrupt your connection to the server while the calculation takes place.

You might notice that the sequences are no longer in fastq format but in [fasta format](https://en.wikipedia.org/wiki/FASTA_format).

The header line (identifier) line, which begins with '>', gives a name and/or a unique identifier for the sequence, and may also contain additional information. Here, we get information about the contigs, their length and k-mer coverage. The remaining lines contain sequence, encoded as nucleotides (ATCG). 

~~~
>contig_1
CACGTTAAATCATATCAGGCGTAATACCACAACCCTTAAGTTAGCGCTTATGGGAATTAT
CCCCGGCTTTTTTATGTATGGTCTTACAGCACCAGTGCTGCGATTGACGCAGACAGCACA
CTCACCAGGGTAGAGCCGTAAACCAGCTTCAGACCGAAGCGAGAAACCACGTTACCTTGC
TCTTCATTCAGGCCTTTAACTGCACCTGCGATAATCCCGATTGAAGAGAAGTTAGCGAAG
GAAACCAGGAACACAGAGATGATGCCTTCAGCACGCGGAGAGAGCGTGGAAGCAATTTTC
TGCAGATCCATCATCGCAACGAACTCGTTGGAAACCAGTTTGGTCGCCATGATACTGCCC
ACTTGCAGTGCTTCACTGGAAGGAACACCCATCACCCATGCAATCGGATAGAAGATGTAG
CCCAGGATGCCCTGGAAGGAGATGCTGTAGCCAAACCAGCCAGTAACGGTGGCAAACAGT
GCGTTCAGCGCGGCGATCAGGGCGATAAAGCCAATCAGCATCGCGGCAACGATAATGGCA
ACTTTGAAACCTGCCAGAATGTATTCACCCAGCATTTCGAAGAAGCTCTGACCTTCGTGC
AGGTTGGACATCTGGATGTTTTCTTCACTGGCATCAACACGGTAAGGATTGATCAGCGAC
AGCACGATAAAGGTGCTGAACATGTTCAGTACCAGCGCAGCAACGACGTATTTCGGTTCC
~~~
{: .bash}


> ## Challenge: How many contigs were generated by SKESA at different k-mer size?
>
> Find out how many contigs there are in the *E. coli* isolate at:
> kmer21
> kmer27
> kmer33
> kmer39
> kmer45
> kmer51
> kmer57
> kmer63
> kmer71
> kmer77
> kmer83
>
> Hint:
> ~~~
> $ grep -c
> ~~~
> prints a count of matching lines for each input file.
> 
> > ## Solution
> >
> > 
> > 
> > `$ grep -c ">kmer21" allkmers.txt `
> > 1256
> > `$ grep -c ">kmer27" allkmers.txt `
> > 602
> >` $ grep -c ">kmer33" allkmers.txt `
> > 470
> > `$ grep -c ">kmer39" allkmers.txt `
> > 384
> >` $ grep -c ">kmer45" allkmers.txt `
> > 287
> >` $ grep -c ">kmer51" allkmers.txt `
> > 256
> >` $ grep -c ">kmer57" allkmers.txt `
> > 238
> >` $ grep -c ">kmer63" allkmers.txt `
> > 222
> >` $ grep -c ">kmer71" allkmers.txt `
> > 216
> >` $ grep -c ">kmer77" allkmers.txt `
> > 208
> >` $ grep -c ">kmer83" allkmers.txt `
> > 198
> > ~~~
> > {: .output}
> {: .solution}
{: .challenge}



{% include links.md %}
