---
title: "Sequence assembly"
exercises: 60
questions:
- "How can the information in the sequencing reads be reduced?"
- "What is the effect of k-mer size on assembly?"
objectives:
- "Assemble the short reads"
keypoints:
- "Assembly is a process which aligns and merges fragments from a longer DNA sequence in order to reconstruct the original sequence."
- "k-mers are short fragments of DNA of length k"
---

Sequence assembly means the alignment and merging of reads in order to reconstruct the original sequence. The assembly of a genome from short sequencing reads will take a while - from minutes up to several hours per genome. 

## Sequence assembly

The assembler we will run is [SPAdes](http://cab.spbu.ru/software/spades/). SPAdes generates a final assembly from multiple kmers. A list of kmers is automatically selected by SPAdes using the maximum read length of the input data, and each individual kmer contributes to the final assembly. If desired, a list of kmers can be specified with the -k flag which will override automatic kmer selection.


To run SPAdes we will use the spades.py command with the --only-assembler option as the reads have already been trimmed, -o for the output folder, -1 for the path to the forward reads, -2 for the path to the reverse reads. To speed up the process, we will be using only two k-mer sizes of 33 and 55 basepair. Normally we would use a range of different k-mers of which SPAdes would select the best result.

~~~
$ cd ~/112018_Secretome/trimmed
$ spades.py -1 ERR022075_1.trimmed.fastq -2 ERR022075_2.trimmed.fastq -k 33,55 -o ~/112018_Secretome/assembly/ERR022075
~~~
{: .bash}

This will take a while. Try not to interrupt your connection to the server while the calculation takes place.

The assemblies for every k-mer setting are found in the folders marked K33 and K55 and are called final_contigs.fasta. The end result before scaffolding is called contigs.fasta and the scaffolded contigs are available in the file called scaffolds.fasta. You might notice that the sequences are no longer in fastq format but in (fasta format)[https://en.wikipedia.org/wiki/FASTA_format].

The header line (identifier) line, which begins with '>', gives a name and/or a unique identifier for the sequence, and may also contain additional information. Here, we get information about the contigs, their length and k-mer coverage. The remaining lines contain sequence, encoded as nucleotides (ATCG). 

~~~
>NODE_1_length_223984_cov_80.658669
CACGTTAAATCATATCAGGCGTAATACCACAACCCTTAAGTTAGCGCTTATGGGAATTAT
CCCCGGCTTTTTTATGTATGGTCTTACAGCACCAGTGCTGCGATTGACGCAGACAGCACA
CTCACCAGGGTAGAGCCGTAAACCAGCTTCAGACCGAAGCGAGAAACCACGTTACCTTGC
TCTTCATTCAGGCCTTTAACTGCACCTGCGATAATCCCGATTGAAGAGAAGTTAGCGAAG
GAAACCAGGAACACAGAGATGATGCCTTCAGCACGCGGAGAGAGCGTGGAAGCAATTTTC
TGCAGATCCATCATCGCAACGAACTCGTTGGAAACCAGTTTGGTCGCCATGATACTGCCC
ACTTGCAGTGCTTCACTGGAAGGAACACCCATCACCCATGCAATCGGATAGAAGATGTAG
CCCAGGATGCCCTGGAAGGAGATGCTGTAGCCAAACCAGCCAGTAACGGTGGCAAACAGT
GCGTTCAGCGCGGCGATCAGGGCGATAAAGCCAATCAGCATCGCGGCAACGATAATGGCA
ACTTTGAAACCTGCCAGAATGTATTCACCCAGCATTTCGAAGAAGCTCTGACCTTCGTGC
AGGTTGGACATCTGGATGTTTTCTTCACTGGCATCAACACGGTAAGGATTGATCAGCGAC
AGCACGATAAAGGTGCTGAACATGTTCAGTACCAGCGCAGCAACGACGTATTTCGGTTCC
~~~
{: .bash}


> ## Challenge: How many contigs were generated by SPAdes?
>
> Find out how many contigs or scaffolds there are in the *E. coli* isolate. 
>
> Hint:
> ~~~
> $ grep -c
> ~~~
> prints a count of matching lines for each input file.
> 
> > ## Solution
> >
> > 
> > ~~~
> > $ grep -c '>'  ERR*/K33/final_contigs.fasta
> > $ grep -c '>'  ERR*/K55/final_contigs.fasta
> > $ grep -c '>'  ERR*/contigs.fasta
> > $ grep -c '>'  ERR*/scaffolds.fasta
> > 
> > ~~~
> > {: .output}
> {: .solution}
{: .challenge}



{% include links.md %}
